CenterForOpenScience/osf.io
# [HotFix] Remove Pending Email for ORCiD Link from Unconfirmed Email on Account Settings Page [#OSF-7549]

## Purpose

### Issue

When existing OSF users link their ORCiD account, they enter a confirmed OSF email to receive the "Link" confirmation email. However, on the "Account Settings" page, this email also appears in the "Unconfirmed Emails" section with an option to "resend confirmation".

First, this is a confirmed email from OSF's perspective which shouldn't be considered as unconfirmed. Second, "resend confirmation" will fail if this happens to be the primary email (most of the time).

### Cause

In the `User` (or `OSFUser` for DjangoOSF), both ORCiD pending confirmation and OSF pending confirmation for adding emails and merging accounts use `.email_verifications`. This `.email_verifications` is further used to calculate the property `.unconfirmed_email`, which is the data source for "Unconfirmed Email" in "Accounts Settings".

```JSON
{  
   "XXXXFhAxCiNslTHqvyzDM5fWItXXXX":{  
      "email":"longze@cos.io",
      "confirmed":false,
      "expiration":{  
         "type":"encoded_datetime",
         "value":"2017-03-15T20:16:41.743218+00:00"
      },
      "external_identity":{  
         "ORCID":{  
            "XXXX-XXXX-XXXX-XXXX":"LINK"
         }
      }
   },
   "XXXXOLZySePe2hKtJqfuMuOSEXXXX":{  
      "email":"longze+a@cos.io",
      "confirmed":false,
      "expiration":{  
         "type":"encoded_datetime",
         "value":"2017-03-15T20:15:48.445343+00:00"
      },
      "external_identity":null
   }
}
```

### Solution

In order to avoid model change during the Django transition, I add a utility method `get_unconfirmed_emails_exclude_external_identity(user)` in `website/profile/utils.py`, Which is only used for KnockOut data-bind.

- Before

![before](https://cloud.githubusercontent.com/assets/3750414/23962664/37d05ce2-0985-11e7-8489-3c436ddd6fd2.png)

- After

![after](https://cloud.githubusercontent.com/assets/3750414/23962679/4321b398-0985-11e7-9f89-903898707ffc.png)

## Changes

### Deployment Notes

HotFix to `master`, and can be then applied to `develop` using `git-flow` without conflicts.

### QA Notes

1. Select one of your OSF accounts
2. (Optional) Go to Settings and unlink ORCiD if it is already linked
3. Login through ORCiD and use the above existing account's primary email to link
4. Do not click the confirmation email sent to you
5. Make sure to log out first (using the "cancel" button or go to osf.io/logout ) to clear current pending session 
6. Go to Settings and confirm that you have a new pending entry in "Connected Identities" and that the email is not in the "Unconfirmed Emails"

In addition, you can test that unconfirmed emails for adding emails or merging accounts still work.

## Side effects

No

## Ticket

https://openscience.atlassian.net/browse/OSF-7549
